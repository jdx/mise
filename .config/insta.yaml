test:
  auto_review: true
review:
  warn_undiscovered: true
