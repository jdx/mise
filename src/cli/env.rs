use eyre::Result;

use crate::cli::args::ToolArg;
use crate::config::Config;
use crate::shell::{get_shell, ShellType};
use crate::toolset::{InstallOptions, Toolset, ToolsetBuilder};

/// Exports env vars to activate mise a single time
///
/// Use this if you don't want to permanently install mise. It's not necessary to
/// use this if you have `mise activate` in your shell rc file.
#[derive(Debug, clap::Args)]
#[clap(visible_alias = "e", verbatim_doc_comment, after_long_help = AFTER_LONG_HELP)]
pub struct Env {
    /// Tool(s) to use
    #[clap(value_name = "TOOL@VERSION")]
    tool: Vec<ToolArg>,

    /// Output in JSON format
    #[clap(long, short = 'J', overrides_with = "shell")]
    json: bool,

    /// Shell type to generate environment variables for
    #[clap(long, short, overrides_with = "json")]
    shell: Option<ShellType>,
}

impl Env {
    pub fn run(self) -> Result<()> {
        let config = Config::try_get()?;
        let mut ts = ToolsetBuilder::new().with_args(&self.tool).build(&config)?;
        ts.install_arg_versions(&config, &InstallOptions::new())?;
        ts.notify_if_versions_missing();

        if self.json {
            self.output_json(&config, ts)
        } else {
            self.output_shell(&config, ts)
        }
    }

    fn output_json(&self, config: &Config, ts: Toolset) -> Result<()> {
        let env = ts.env_with_path(config)?;
        miseprintln!("{}", serde_json::to_string_pretty(&env)?)?;
        Ok(())
    }

    fn output_shell(&self, config: &Config, ts: Toolset) -> Result<()> {
        let default_shell = get_shell(Some(ShellType::Bash)).unwrap();
        let shell = get_shell(self.shell).unwrap_or(default_shell);
        for (k, v) in ts.env_with_path(config)? {
            let k = k.to_string();
            let v = v.to_string();
            miseprint!("{}", shell.set_env(&k, &v))?;
        }
        Ok(())
    }
}

static AFTER_LONG_HELP: &str = color_print::cstr!(
    r#"<bold><underline>Examples:</underline></bold>
  $ <bold>eval "$(mise env -s bash)"</bold>
  $ <bold>eval "$(mise env -s zsh)"</bold>
  $ <bold>mise env -s fish | source</bold>
  $ <bold>execx($(mise env -s xonsh))</bold>
"#
);

#[cfg(test)]
mod tests {
    use crate::cli::tests::grep;
    use std::env;

    use crate::dirs;

    #[test]
    fn test_env() {
        let stdout = assert_cli!("env", "-s", "bash");
        assert!(stdout.contains(
            dirs::DATA
                .join("installs/tiny/3/bin")
                .to_string_lossy()
                .as_ref()
        ));
    }

    #[test]
    fn test_env_with_runtime_arg() {
        assert_cli!("install", "tiny@3.0");
        let stdout = assert_cli!("env", "tiny@3.0", "-s", "bash");

        assert!(stdout.contains(
            dirs::DATA
                .join("installs/tiny/3.0/bin")
                .to_string_lossy()
                .as_ref()
        ));
    }

    #[test]
    fn test_env_alias() {
        assert_cli!("plugin", "add", "tiny");
        assert_cli!("install", "tiny@my/alias");
        let stdout = assert_cli!("env", "tiny@my/alias", "-s", "bash");
        assert!(stdout.contains(
            dirs::DATA
                .join("installs/tiny/3.0.1")
                .to_string_lossy()
                .as_ref()
        ));
    }

    #[test]
    fn test_env_tiny() {
        let stdout = assert_cli!("env", "tiny@2", "tiny@1", "tiny@3", "-s", "bash");
        assert_str_eq!(grep(stdout, "JDXCODE"), "export JDXCODE_TINY=2.1.0");
    }

    #[test]
    fn test_env_default_shell() {
        env::set_var("SHELL", "");
        let stdout = assert_cli!("env");
        assert!(stdout.contains("export PATH="));
    }

    #[test]
    fn test_env_json() {
        assert_cli_snapshot!("env", "-J");
    }
}
