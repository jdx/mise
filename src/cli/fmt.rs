use crate::config::ALL_TOML_CONFIG_FILES;
use crate::Result;
use crate::{config, dirs, file};
use eyre::bail;
use taplo::formatter::Options;

/// Formats mise.toml
///
/// Sorts keys and cleans up whitespace in mise.toml
#[derive(Debug, clap::Args)]
#[clap(verbatim_doc_comment, after_long_help = AFTER_LONG_HELP)]
pub struct Fmt {
    /// Format all files from the current directory
    #[clap(short, long)]
    pub all: bool,
}

impl Fmt {
    pub fn run(self) -> eyre::Result<()> {
        let cwd = dirs::CWD.clone().unwrap_or_default();
        let configs = if self.all {
            ALL_TOML_CONFIG_FILES.clone()
        } else {
            config::config_files_in_dir(&cwd)
        };
        if configs.is_empty() {
            bail!("No config file found in current directory");
        }
        for p in configs {
            if !p
                .file_name()
                .is_some_and(|f| f.to_string_lossy().ends_with("toml"))
            {
                continue;
            }
            let toml = file::read_to_string(&p)?;
            let toml = sort(toml)?;
            let toml = taplo::formatter::format(
                &toml,
                Options {
                    align_entries: false,
                    align_comments: true,
                    align_single_comments: true,
                    array_trailing_comma: true,
                    array_auto_expand: true,
                    inline_table_expand: true,
                    array_auto_collapse: true,
                    compact_arrays: true,
                    compact_inline_tables: false,
                    compact_entries: false,
                    column_width: 80,
                    indent_tables: false,
                    indent_entries: false,
                    indent_string: "  ".to_string(),
                    trailing_newline: true,
                    reorder_keys: false,
                    reorder_arrays: false,
                    allowed_blank_lines: 2,
                    crlf: false,
                },
            );
            file::write(&p, &toml)?;
        }

        Ok(())
    }
}

fn sort(toml: String) -> Result<String> {
    let mut doc: toml_edit::DocumentMut = toml.parse()?;
    let order = |k: String| match k.as_str() {
        "min_version" => 0,
        "env_file" => 1,
        "env_path" => 2,
        "_" => 3,
        "env" => 4,
        "vars" => 5,
        "hooks" => 6,
        "watch_files" => 7,
        "tools" => 8,
        "tasks" => 10,
        "task_config" => 11,
        "redactions" => 12,
        "alias" => 13,
        "plugins" => 14,
        "settings" => 15,
        _ => 9,
    };
    doc.sort_values_by(|a, _, b, _| order(a.to_string()).cmp(&order(b.to_string())));
    Ok(doc.to_string())
}

static AFTER_LONG_HELP: &str = color_print::cstr!(
    r#"<bold><underline>Examples:</underline></bold>

    $ <bold>mise fmt</bold>
"#
);
