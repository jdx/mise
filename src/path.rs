pub use std::path::*;

use crate::dirs;

pub trait PathExt {
    /// replaces $HOME with "~"
    fn display_user(&self) -> String;
    fn mount(&self, on: &Path) -> PathBuf;
    fn is_empty(&self) -> bool;
}

impl PathExt for Path {
    fn display_user(&self) -> String {
        let home = dirs::HOME.to_string_lossy();
        match cfg!(unix) && self.starts_with(home.as_ref()) && home != "/" {
            true => self.to_string_lossy().replacen(home.as_ref(), "~", 1),
            false => self.to_string_lossy().to_string(),
        }
    }

    fn mount(&self, on: &Path) -> PathBuf {
        if self.is_empty() {
            on.to_path_buf()
        } else {
            on.join(self)
        }
    }

    fn is_empty(&self) -> bool {
        self.as_os_str().is_empty()
    }
}
