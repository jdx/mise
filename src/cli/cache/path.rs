use eyre::Result;

use crate::env;

/// Show the cache directory path
#[derive(Debug, clap::Args)]
#[clap(verbatim_doc_comment, visible_alias = "dir")]
pub struct CachePath {}

impl CachePath {
    pub fn run(self) -> Result<()> {
        miseprintln!("{}", env::MISE_CACHE_DIR.display());
        Ok(())
    }
}
