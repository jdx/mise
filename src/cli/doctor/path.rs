use crate::config::Config;
use crate::Result;
use std::env;

/// Print the current PATH entries mise is providing
#[derive(Debug, clap::Args)]
#[clap(alias="paths", verbatim_doc_comment, after_long_help = AFTER_LONG_HELP)]
pub struct Path {
    /// Print all entries including those not provided by mise
    #[clap(long, short, verbatim_doc_comment)]
    full: bool,
}

impl Path {
    pub fn run(self) -> Result<()> {
        let config = Config::get();
        let ts = config.get_toolset()?;
        let paths = if self.full {
            let env = ts.env_with_path(&config)?;
            let path = env.get("PATH").cloned().unwrap_or_default();
            env::split_paths(&path).collect()
        } else {
            ts.list_final_paths()?
        };
        for path in paths {
            println!("{}", path.display());
        }
        Ok(())
    }
}

static AFTER_LONG_HELP: &str = color_print::cstr!(
    r#"<bold><underline>Examples:</underline></bold>

    Get the current PATH entries mise is providing
    $ mise path
    /home/user/.local/share/mise/installs/node/24.0.0/bin
    /home/user/.local/share/mise/installs/rust/1.90.0/bin
    /home/user/.local/share/mise/installs/python/3.10.0/bin
"#
);
