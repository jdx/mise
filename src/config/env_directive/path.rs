use crate::config::env_directive::{EnvResults, PathEntry};
use crate::result;
use std::path::PathBuf;

impl EnvResults {
    pub fn path(
        ctx: &mut tera::Context,
        r: &mut EnvResults,
        paths: &mut Vec<(PathEntry, PathBuf)>,
        source: PathBuf,
        input_str: PathEntry,
    ) -> result::Result<()> {
        // trace!("resolve: input_str: {:#?}", input_str);
        match input_str {
            PathEntry::Normal(input) => {
                // trace!(
                //     "resolve: normal: input: {:?}, input.to_string(): {:?}",
                //     &input,
                //     input.to_string_lossy().as_ref()
                // );
                let s = r.parse_template(ctx, &source, input.to_string_lossy().as_ref())?;
                // trace!("resolve: s: {:?}", &s);
                paths.push((PathEntry::Normal(s.into()), source));
            }
            PathEntry::Lazy(input) => {
                // trace!(
                //     "resolve: lazy: input: {:?}, input.to_string(): {:?}",
                //     &input,
                //     input.to_string_lossy().as_ref()
                // );
                paths.push((PathEntry::Lazy(input), source));
            }
        }
        Ok(())
    }
}

#[cfg(test)]
#[cfg(unix)]
mod tests {
    use super::*;
    use crate::config::env_directive::EnvDirective;
    use crate::env_diff::EnvMap;
    use crate::tera::BASE_CONTEXT;
    use crate::test::replace_path;
    use insta::assert_debug_snapshot;
    use test_log::test;

    #[test]
    fn test_env_path() {
        let mut env = EnvMap::new();
        env.insert("A".to_string(), "1".to_string());
        env.insert("B".to_string(), "2".to_string());
        let results = EnvResults::resolve(
            BASE_CONTEXT.clone(),
            &env,
            vec![
                (
                    EnvDirective::Path("/path/1".into()),
                    PathBuf::from("/config"),
                ),
                (
                    EnvDirective::Path("/path/2".into()),
                    PathBuf::from("/config"),
                ),
                (
                    EnvDirective::Path("~/foo/{{ env.A }}".into()),
                    Default::default(),
                ),
                (
                    EnvDirective::Path("./rel/{{ env.A }}:./rel2/{{env.B}}".into()),
                    Default::default(),
                ),
            ],
        )
        .unwrap();
        assert_debug_snapshot!(
            results.env_paths.into_iter().map(|p| replace_path(&p.display().to_string())).collect::<Vec<_>>(),
            @r#"
        [
            "/path/1",
            "/path/2",
            "~/foo/1",
            "~/cwd/rel/1",
            "~/cwd/rel2/2",
        ]
        "#
        );
    }
}
