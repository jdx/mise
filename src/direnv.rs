use base64::prelude::*;
use color_eyre::eyre::Result;
use flate2::write::{ZlibDecoder, ZlibEncoder};
use flate2::Compression;
use itertools::Itertools;
use serde_derive::{Deserialize, Serialize};
use std::collections::HashMap;
use std::env::{join_paths, split_paths};
use std::fmt::{Display, Formatter};
use std::io::Write;

use std::path::{Path, PathBuf};

#[derive(Debug, Serialize, Deserialize)]
pub struct DirenvDiff {
    #[serde(default, rename = "p")]
    pub old: HashMap<String, String>,
    #[serde(default, rename = "n")]
    pub new: HashMap<String, String>,
}

impl DirenvDiff {
    pub fn parse(input: &str) -> Result<DirenvDiff> {
        // let bytes = BASE64_URL_SAFE.decode(input)?;
        // let uncompressed = inflate_bytes_zlib(&bytes).unwrap();
        // Ok(serde_json::from_slice(&uncompressed[..])?)
        let mut writer = Vec::new();
        let mut decoder = ZlibDecoder::new(writer);
        let bytes = BASE64_URL_SAFE.decode(input)?;
        decoder.write_all(&bytes[..])?;
        writer = decoder.finish()?;
        Ok(serde_json::from_slice(&writer[..])?)
    }

    pub fn new_path(&self) -> Vec<PathBuf> {
        let path = self.new.get("PATH");
        match path {
            Some(path) => split_paths(path).collect(),
            None => vec![],
        }
    }

    pub fn old_path(&self) -> Vec<PathBuf> {
        let path = self.old.get("PATH");
        match path {
            Some(path) => split_paths(path).collect(),
            None => vec![],
        }
    }

    /// this adds a directory to both the old and new path in DIRENV_DIFF
    /// the purpose is to trick direnv into thinking that this path has always been there
    /// that way it does not remove it when it modifies PATH
    /// it returns the old and new paths as vectors
    pub fn add_path_to_old_and_new(&mut self, path: &Path) -> Result<(Vec<PathBuf>, Vec<PathBuf>)> {
        let mut old = self.old_path();
        let mut new = self.new_path();

        old.insert(0, path.into());
        new.insert(0, path.into());

        self.old
            .insert("PATH".into(), join_paths(&old)?.into_string().unwrap());
        self.new
            .insert("PATH".into(), join_paths(&new)?.into_string().unwrap());

        Ok((old, new))
    }

    pub fn remove_path_from_old_and_new(
        &mut self,
        path: &Path,
    ) -> Result<(Vec<PathBuf>, Vec<PathBuf>)> {
        let mut old = self.old_path();
        let mut new = self.new_path();

        // remove the path from both old and new but only once
        old.iter().position(|p| p == path).map(|i| old.remove(i));
        new.iter().position(|p| p == path).map(|i| new.remove(i));

        self.old
            .insert("PATH".into(), join_paths(&old)?.into_string().unwrap());
        self.new
            .insert("PATH".into(), join_paths(&new)?.into_string().unwrap());

        Ok((old, new))
    }

    pub fn dump(&self) -> Result<String> {
        let mut gz = ZlibEncoder::new(Vec::new(), Compression::fast());
        gz.write_all(&serde_json::to_vec(self)?)?;
        Ok(BASE64_URL_SAFE.encode(gz.finish()?))
    }
}

impl Display for DirenvDiff {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        let print_sorted = |hashmap: &HashMap<String, String>| {
            hashmap
                .iter()
                .map(|(k, v)| format!("{k}={v}"))
                .sorted()
                .collect::<Vec<_>>()
        };
        f.debug_struct("DirenvDiff")
            .field("old", &print_sorted(&self.old))
            .field("new", &print_sorted(&self.new))
            .finish()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_display_snapshot;

    #[test]
    fn test_parse() {
        let input = r#"eJys0c1yojAAwPF3ybmWaLB-zPSAGCqIQCGgeGGIELDlM2BEOr77zs7szr7AXv-H3-X_Axqw_gGabYM1qPk1A88XUP1OW93FVhBtdReswURq-FXEfSqJmEusLpKUdxLspALRJY1Yt2Bifk8aLhf5iiZIhhDCjEtE6svmteGuSJVHAV7-qppuYrAG_0WVXtNK8Ms__KgQdYc9sAapMXRj1-9XW8VX7A16UA4NPIs9xCK5WO51XnvfwWBT1R9N7zIcHvvJbZF5g8pk0V2c5CboIw8_NjOUWDK5qcxIcaFrp3anhwdr5FeKJmfd9stgqvuVZqcXsXHYJ-kSGWpoxyZLzf0a0LUcMgv17exenXXunfOTZZfybiVmb9OAhjDtHEcOk0lrRWG84OrRobW6IgGGZqwelglTq8UmJrbP9p0x9pTW5t3L21P1mZfL7_pMtIW599v-Cx_dmzEdCcZ1TAzkz7dvfO4QAefO6Y4VxYmijzgP_Oz9Hbz8uU5jDp7PXwEAAP__wB6qKg=="#;
        let diff = DirenvDiff::parse(input).unwrap();
        assert_display_snapshot!(diff);
    }

    #[test]
    fn test_dump() {
        let diff = DirenvDiff {
            old: HashMap::from([("a".to_string(), "b".to_string())]),
            new: HashMap::from([("c".to_string(), "d".to_string())]),
        };
        let output = diff.dump().unwrap();
        assert_display_snapshot!(&output);
        let diff = DirenvDiff::parse(&output).unwrap();
        assert_display_snapshot!(diff);
    }

    #[test]
    fn test_add_path_to_old_and_new() {
        let mut diff = DirenvDiff {
            old: HashMap::from([("PATH".to_string(), "/foo:/tmp:/bar:/old".to_string())]),
            new: HashMap::from([("PATH".to_string(), "/foo:/bar:/new".to_string())]),
        };
        let path = PathBuf::from("/tmp");
        diff.add_path_to_old_and_new(&path).unwrap();
        assert_display_snapshot!(diff.old.get("PATH").unwrap());
        assert_display_snapshot!(diff.new.get("PATH").unwrap());
    }

    #[test]
    fn test_null_path() {
        let mut diff = DirenvDiff {
            old: HashMap::from([]),
            new: HashMap::from([]),
        };
        let path = PathBuf::from("/tmp");
        diff.add_path_to_old_and_new(&path).unwrap();
        assert_display_snapshot!(diff.old.get("PATH").unwrap());
        assert_display_snapshot!(diff.new.get("PATH").unwrap());
    }
}
