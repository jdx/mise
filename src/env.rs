use std::collections::{HashMap, HashSet};
pub use std::env::*;
use std::path::PathBuf;

use itertools::Itertools;
use lazy_static::lazy_static;
use log::LevelFilter;

use crate::env_diff::{EnvDiff, EnvDiffOperation, EnvDiffPatches};

lazy_static! {
    pub static ref ARGS: Vec<String> = args().collect();
    pub static ref HOME: PathBuf = if cfg!(test) {
        PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("test")
    } else {
        dirs_next::home_dir().unwrap_or_else(|| PathBuf::from("/"))
    };
    pub static ref PWD: PathBuf = if cfg!(test) {
        HOME.join("cwd")
    } else {
        current_dir().unwrap_or_else(|_| PathBuf::new())
    };
    pub static ref XDG_CACHE_HOME: PathBuf = if cfg!(test) {
        HOME.join("cache")
    } else {
        var_os("XDG_CACHE_HOME")
            .map(PathBuf::from)
            .unwrap_or_else(|| HOME.join(".cache"))
    };
    pub static ref XDG_DATA_HOME: PathBuf = if cfg!(test) {
        HOME.join("data")
    } else {
        var_os("XDG_DATA_HOME")
            .map(PathBuf::from)
            .unwrap_or_else(|| HOME.join(".local/share"))
    };
    pub static ref XDG_CONFIG_HOME: PathBuf = if cfg!(test) {
        HOME.join("config")
    } else {
        var_os("XDG_CONFIG_HOME")
            .map(PathBuf::from)
            .unwrap_or_else(|| HOME.join(".config"))
    };
    pub static ref RTX_CACHE_DIR: PathBuf = if cfg!(test) {
        XDG_CACHE_HOME.clone()
    } else {
        var_os("RTX_CACHE_DIR")
            .map(PathBuf::from)
            .unwrap_or_else(|| XDG_CACHE_HOME.join("rtx"))
    };
    pub static ref RTX_CONFIG_DIR: PathBuf = if cfg!(test) {
        XDG_CONFIG_HOME.clone()
    } else {
        var_os("RTX_CONFIG_DIR")
            .map(PathBuf::from)
            .unwrap_or_else(|| XDG_CONFIG_HOME.join("rtx"))
    };
    pub static ref RTX_DATA_DIR: PathBuf = if cfg!(test) {
        XDG_DATA_HOME.clone()
    } else {
        var_os("RTX_DATA_DIR")
            .map(PathBuf::from)
            .unwrap_or_else(|| XDG_DATA_HOME.join("rtx"))
    };
    pub static ref RTX_TMP_DIR: PathBuf = temp_dir().join("rtx");
    pub static ref SHELL: String = var("SHELL").unwrap_or_else(|_| "sh".into());
    pub static ref RTX_EXE: PathBuf = current_exe().unwrap_or_else(|_| "rtx".into());
    pub static ref RTX_LOG_LEVEL: log::LevelFilter = {
        let log_level = var("RTX_LOG_LEVEL").unwrap_or_default();
        match log_level.parse::<LevelFilter>() {
            Ok(level) => level,
            _ => {
                if *RTX_TRACE {
                    log::LevelFilter::Trace
                } else if *RTX_DEBUG {
                    log::LevelFilter::Debug
                } else if *RTX_QUIET {
                    log::LevelFilter::Warn
                } else {
                    log::LevelFilter::Info
                }
            }
        }
    };
    pub static ref RTX_LOG_FILE_LEVEL: log::LevelFilter = {
        let log_level = var("RTX_LOG_FILE_LEVEL").unwrap_or_default();
        match log_level.parse::<log::LevelFilter>() {
            Ok(level) => level,
            _ => *RTX_LOG_LEVEL,
        }
    };
    pub static ref RTX_MISSING_RUNTIME_BEHAVIOR: Option<String> = if cfg!(test) {
        Some("autoinstall".into())
    } else {
        var("RTX_MISSING_RUNTIME_BEHAVIOR").ok()
    };
    pub static ref __RTX_DIFF: EnvDiff = get_env_diff();
    pub static ref RTX_QUIET: bool = var_is_true("RTX_QUIET");
    pub static ref RTX_DEBUG: bool = var_is_true("RTX_DEBUG");
    pub static ref RTX_TRACE: bool = var_is_true("RTX_TRACE");
    pub static ref RTX_VERBOSE: bool = *RTX_DEBUG || *RTX_TRACE || var_is_true("RTX_VERBOSE");
    pub static ref RTX_JOBS: usize = var("RTX_JOBS")
        .ok()
        .and_then(|v| v.parse::<usize>().ok())
        .unwrap_or(4);
    /// essentially, this is whether we show spinners or build output on runtime install
    pub static ref PRISTINE_ENV: HashMap<String, String> =
        get_pristine_env(&__RTX_DIFF, vars().collect());
    pub static ref PATH: Vec<PathBuf> = match PRISTINE_ENV.get("PATH") {
        Some(path) => split_paths(path).collect(),
        None => vec![],
    };
    pub static ref RTX_DEFAULT_TOOL_VERSIONS_FILENAME: String = if cfg!(test) {
        ".tool-versions".into()
    } else {
        var("RTX_DEFAULT_TOOL_VERSIONS_FILENAME").unwrap_or_else(|_| ".tool-versions".into())
    };
    pub static ref DIRENV_DIR: Option<String> = var("DIRENV_DIR").ok();
    pub static ref DIRENV_DIFF: Option<String> = var("DIRENV_DIFF").ok();
    pub static ref RTX_HIDE_OUTDATED_BUILD: bool = var_is_true("RTX_HIDE_OUTDATED_BUILD");
}

fn get_env_diff() -> EnvDiff {
    let env = vars().collect::<HashMap<_, _>>();
    match env.get("__RTX_DIFF") {
        Some(raw) => EnvDiff::deserialize(raw).unwrap_or_else(|err| {
            warn!("Failed to deserialize __RTX_DIFF: {}", err);
            EnvDiff::default()
        }),
        None => EnvDiff::default(),
    }
}

fn var_is_true(key: &str) -> bool {
    match var(key) {
        Ok(v) => {
            let v = v.to_lowercase();
            !v.is_empty()
                && v != "n"
                && v != "no"
                && v != "false"
                && v != "0"
                && v != "off"
                && v != " "
        }
        Err(_) => false,
    }
}

/// this returns the environment as if __RTX_DIFF was reversed.
/// putting the shell back into a state before hook-env was run
fn get_pristine_env(
    rtx_diff: &EnvDiff,
    orig_env: HashMap<String, String>,
) -> HashMap<String, String> {
    let patches = rtx_diff.reverse().to_patches();
    let mut env = apply_patches(&orig_env, &patches);

    // get the current path as a vector
    let path = match env.get("PATH") {
        Some(path) => split_paths(path).collect(),
        None => vec![],
    };
    // get the paths that were removed by rtx as a hashset
    let mut to_remove = rtx_diff.path.iter().collect::<HashSet<_>>();

    // remove those paths that were added by rtx, but only once (the first time)
    let path = path
        .into_iter()
        .filter(|p| !to_remove.remove(p))
        .collect_vec();

    // put the pristine PATH back into the environment
    env.insert(
        "PATH".into(),
        join_paths(path).unwrap().to_string_lossy().to_string(),
    );
    env
}

fn apply_patches(
    env: &HashMap<String, String>,
    patches: &EnvDiffPatches,
) -> HashMap<String, String> {
    let mut new_env = env.clone();
    for patch in patches {
        match patch {
            EnvDiffOperation::Add(k, v) | EnvDiffOperation::Change(k, v) => {
                new_env.insert(k.into(), v.into());
            }
            EnvDiffOperation::Remove(k) => {
                new_env.remove(k);
            }
        }
    }

    new_env
}

#[test]
fn test_apply_patches() {
    let mut env = HashMap::new();
    env.insert("foo".into(), "bar".into());
    env.insert("baz".into(), "qux".into());
    let patches = vec![
        EnvDiffOperation::Add("foo".into(), "bar".into()),
        EnvDiffOperation::Change("baz".into(), "qux".into()),
        EnvDiffOperation::Remove("quux".into()),
    ];
    let new_env = apply_patches(&env, &patches);
    assert_eq!(new_env.len(), 2);
    assert_eq!(new_env.get("foo").unwrap(), "bar");
    assert_eq!(new_env.get("baz").unwrap(), "qux");
}
