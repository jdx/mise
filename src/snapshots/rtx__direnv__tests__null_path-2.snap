---
source: src/direnv.rs
expression: "diff.new.get(\"PATH\").unwrap()"
---
/tmp
